### PR TITLE
Bump master bootstrap compiler

### PR DESCRIPTION
**Pull request by emilyalbini**
This PR bumps the bootstrap compiler to the beta created earlier this week, cherry-picks the stabilization version number updates, and updates the `cfg(bootstrap)`s.

r? @Mark-Simulacrum

### CODE REVIEW

File: library/core/src/future/mod.rs [hunk 1]
```diff
@@ -62,7 +62,7 @@ unsafe impl Sync for ResumeTy {}
 /// better error messages (`impl Future` rather than `GenFuture<[closure.....]>`).
 // This is `const` to avoid extra errors after we recover from `const async fn`
 // FIXME(swatinem): This fn can be removed when bumping the bootstrap compiler
```
Comment (Nilstrieb) [inline library/core/src/future/mod.rs:64]: @Swatinem can the entire function be removed now?

Comment (pietroalbini) [reply library/core/src/future/mod.rs:64]: Removed the unknown lang items for now, IMO it's fine to merge this if it builds and cleanup the functions (if needed) later.

Comment (Swatinem) [reply library/core/src/future/mod.rs:64]: Yes, the idea was that this whole function can be removed now, as it is now unused.

Comment (Mark-Simulacrum) [reply library/core/src/future/mod.rs:64]: Happy to see a separate PR with more expansive cleanup, but I think minimal work as part of this PR is best.

**Comment by Mark-Simulacrum** [comment]
r=me modulo CI

File: library/std/src/sys/unix/weak.rs [hunk 2]
```diff
@@ -43,7 +43,7 @@ pub(crate) macro weak {
     )
 }
 
-#[cfg(all(not(any(target_os = "macos", target_os = "ios")), bootstrap))]
+#[cfg(not(any(target_os = "macos", target_os = "ios")))]
 pub(crate) macro weak {
```
Comment (cuviper) [inline library/std/src/sys/unix/weak.rs:47]: This whole item should go away with `bootstrap`, in favor of the `not(bootstrap)` one above.

File: library/core/src/intrinsics.rs [hunk 3]
```diff
@@ -55,11 +55,9 @@
 #![allow(missing_docs)]
 
 use crate::marker::DiscriminantKind;
-#[cfg(not(bootstrap))]
 use crate::marker::Tuple;
 use crate::mem;
 
-#[cfg(not(bootstrap))]
```
Comment (JakobDegen) [inline library/core/src/intrinsics.rs:62]: The doctests in here require #105356 , which I believe is not yet included in the bootstrap compiler this is using. This particular bump can be fixed by not removing this  just yet. I'll add more granular `#[cfg(not(bootstrap))]` in the future, which should prevent this from reoccuring

**Comment by jyn514** [comment]
I think this will need https://github.com/rust-lang/rust/pull/105899/ too to avoid rustdoc failing - @lukas-code can you please put together a patch based on this PR without the `cfg(bootstrap)`?

**Comment by lukas-code** [comment]
> can you please put together a patch based on this PR without the `cfg(bootstrap)`?

I've prepared a patch at https://github.com/lukas-code/rust/pull/4, but it's still blocked on https://github.com/rust-lang/rust/pull/105915 getting backported.

**Comment by lqd** [comment]
Apparently https://github.com/rust-lang/rust/pull/105800 is still not enough to make `rust.lto = thin` work at stage0. Maybe we can take care of this later and not remove this [`cfg(bootstrap)`](https://github.com/rust-lang/rust/pull/105741/commits/fdf6cc34b2564100c5d07b573ef246124ee45b47#diff-cc65477e7704ede82e44907dcff240ecb0a547eddfadc515bb706e724068f8feL694) for now ?

**Comment by lukas-code** [comment]
Stage0 LTO seems to be fixed with this patch. I have no idea if that makes sense though, and it might be better to just disable stage0 LTO, because it takes forever for little benefit.

```patch
diff --git a/src/bootstrap/compile.rs b/src/bootstrap/compile.rs
index ae5dfdea19f..40c4fcfd984 100644
--- a/src/bootstrap/compile.rs
+++ b/src/bootstrap/compile.rs
@@ -127,6 +127,13 @@ fn run(self, builder: &Builder<'_>) {
         let mut cargo = builder.cargo(compiler, Mode::Std, SourceType::InTree, target, "build");
         std_cargo(builder, target, compiler.stage, &mut cargo);
 
+        match builder.config.rust_lto {
+            RustcLto::Thin | RustcLto::Fat => {
+                cargo.rustflag("-Cembed-bitcode=yes");
+            }
+            RustcLto::ThinLocal => {}
+        }
+
         builder.info(&format!(
             "Building stage{} std artifacts ({} -> {})",
             compiler.stage, &compiler.host, target
```

**Comment by Mark-Simulacrum** [comment]
Restored the gate on stage 0, but also dropped an explicit cfg(bootstrap) in the comment. I'm not sure we actually want cfg(bootstrap) there -- doing LTO in stage 0 doesn't feel very helpful.

@bors r+ rollup=never

**Comment by Mark-Simulacrum** [comment]
These benchmarks are historically prone to noise and this change is required regardless.

**Comment by lqd** [comment]
coercions has even already reverted to its steady-state

image

**Pull request merged**